monkey/monkey_island/cc/services/authentication_service/setup.py

```diff
@@ -50,3 +54,32 @@ def _build_authentication_facade(container: DIContainer, security: Security):
         token_parser,
         container.resolve(MongoOTPRepository),
     )
+
+
+def _register_event_handlers(container: DIContainer, authentication_facade: AuthenticationFacade):
+    agent_event_queue = container.resolve(IAgentEventQueue)
+    island_event_queue = container.resolve(IIslandEventQueue)
+
+    agent_event_queue.subscribe_type(
+        AgentShutdownEvent, unregister_agent_on_shutdown(authentication_facade)
+    )
+    island_event_queue.subscribe(
+        IslandEventTopic.AGENT_TIMED_OUT, unregister_agent_on_timeout(authentication_facade)
+    )
+
+
+class unregister_agent_on_shutdown:
+    def __init__(self, authentication_facade: AuthenticationFacade):
+        self._authentication_facade = authentication_facade
+
+    def __call__(self, event: AbstractAgentEvent):
+        agent_id = event.source
+        self._authentication_facade.remove_user(str(agent_id))
+
+
+class unregister_agent_on_timeout:
+    def __init__(self, authentication_facade: AuthenticationFacade):
+        self._authentication_facade = authentication_facade
+
+    def __call__(self, agent_id: AgentID):
+        self._authentication_facade.remove_user(str(agent_id))
```

Comment:
This feels out of place in the authentication service, and I think all event handling and subscription should be isolated to where the rest of it is done. 

What if we moved the logic in `unregister_agent_on_shutdown` to `monkey_island/cc/agent_event_handlers/update_agent_shutdown_status.py`, and the logic in `unregister_agent_on_timeout` to `AgentHeartbeatMonitor.set_unresponsive_agents_stop_time()`?

Alternatively, if we absolutely want to use pubsub for this, we could have a separate function for registering these events in `island_event_handlers.py` that is called after authentication is setup.

Reply:
`AuthenticationFacade` is not exposed outside of the authentication service, nor should it be. The authentication service encapsulates all authentication-related code. It is highly cohesive. In contrast, the `monkey_island/cc/agent_event_handlers` is only slightly cohesive. That is, the components in that package have in common that they subscribe to events, but other than that they may be totally unrelated.

Moving details about unregistering agents into `AgentHeartbeatMonitor` is undesirable because:
1. It reduces the cohesion of `AgentHeartbeatMonitor`. Whereas now it is only responsible handling agent timeouts, moving authentication logic into it would also make it responsible for some component of the authentication system.
2. It makes the authentication system less cohesive as a whole. It spreads some business logic out into an unrelated component. This makes it difficult to reason about the behavior of the authentication system, since you'll need to audit the whole code base to understand all of the different, unrelated components that affect how the authentication system works.

All of the authentication logic should be consolidated within the authentication service. If the authentication service is highly cohesive and comprehensive, you can go to exactly one place to understand how authentication works.